server/src/main/java/org/elasticsearch/index/mapper/SourceValueFetcher.java

```diff
@@ -39,17 +39,19 @@
     private final Set<String> sourcePaths;
     private final @Nullable Object nullValue;
 
-    public SourceValueFetcher(String fieldName, MapperService mapperService) {
-        this(fieldName, mapperService, null);
+    /**
+     * @param sourcePaths The paths in the source to read.
+     */
+    public SourceValueFetcher(Set<String> sourcePaths) {
+        this(sourcePaths, null);
     }
 
     /**
-     * @param fieldName The name of the field.
-     * @param mapperService A mapper service.
+     * @param sourcePaths The paths in the source to read.
      * @param nullValue A optional substitute value if the _source value is 'null'.
      */
-    public SourceValueFetcher(String fieldName, MapperService mapperService, Object nullValue) {
-        this.sourcePaths = mapperService.sourcePath(fieldName);
+    public SourceValueFetcher(Set<String> sourcePaths, @Nullable Object nullValue) {
```

Comment:
I think that removing the MapperService dependency from SourceValueFetcher should be a separate change. It is the greater part of this diff and that will make reviewing the actual change (mapping snapshot) easier

Reply:
++